src/main/scala/calc/Compiler.scala

```diff
@@ -61,12 +64,34 @@ object Compiler {
     ir.Hashers.hashClassDef(classDef)
   }
 
+
+
+  def findCapture(tree: Tree): Set[String] = {
+    tree match {
+      case _:Literal => Set()
+      case BinaryOp(_, l, r) => findCapture(l) ++ findCapture(r)
+      case Let(i, v, body) => findCapture(body) ++ findCapture(v) - i.name
+      case Ident(n) => Set(n)
+      case If(cond, thenp, elsep) => findCapture(cond) ++ findCapture(thenp) ++ findCapture(elsep)
+      case Closure(idents, body) => findCapture(body) -- idents.toSet.flatMap((x:Tree) => findCapture(x))
+      case Call(f, args) => findCapture(f) ++ args.toSet.flatMap((x:Tree) => findCapture(x))
+    }
+  }
+
+  def compileClosure(tree: Closure, ident: String)(implicit env:Map[String, irt.Tree],typeEnv:Map[String, irtpe.Type]): irt.Tree ={
+    val captureSet = findCapture(tree) - ident
+    val captureParam = captureSet.map(s => irt.ParamDef(irt.Ident(s)(tree.pos), typeEnv(s), false, false)(tree.pos)).toList
+    val params = tree.params.map(s => irt.ParamDef(irt.Ident(s.name)(tree.pos), irtpe.AnyType, false, false)(tree.pos))
+    irt.Closure(captureParam, params, compileExpr(tree.body)(env, typeEnv), captureSet.map(s => env(s)).toList)(tree.pos)
+  }
+
+
   /** Compile an expression tree into an IR `Tree`, which is an expression
    *  that evaluates to the result of the tree.
    *
    *  This is the main method you have to implement.
    */
-  def compileExpr(tree: Tree): irt.Tree = {
+  def compileExpr(tree: Tree)(implicit env:Map[String, irt.Tree],typeEnv:Map[String, irtpe.Type]): irt.Tree = {
```

Comment:
Rather than using implicit parameters, I recommend define a private overload of `compileExpr` (or even with a different name). It is not good practice to use implicit parameters of "general" types (like `Map`).